### PR TITLE
[Notifications Refresh] Comment Content Redesign

### PR DESCRIPTION
**Pull request by salimbraksa**
Fixes https://github.com/Automattic/wordpress-mobile/issues/31 and part of https://github.com/Automattic/wordpress-mobile/issues/38

## Dependencies
This PR depends on:
- https://github.com/wordpress-mobile/WordPress-iOS/pull/23117

## Description

This PR introduces a new component for the comment table view cell, while the previous cell, `CommentContentTableViewCell`, continues to be used in the Reader Comments screen. I decided against refactoring the original for a couple of reasons:
1. There's a risk of causing regression issues specifically on the Reader Comments screen.
2. The existing class is already quite complex and lengthy. Adding further logic to manage styling would complicate it even more.

The new class `CommentDetailContentTableViewCell` is comprised of 2 components:
- `CommentContentHeaderView`:  This component was implemented in this [PR](https://github.com/pulls).
- `CommentDetailContentView`: I've extracted the comment rendering logic from `CommentContentTableViewCell` into this component.

| Comment | Menu |
| --------- | ------ |
|  |  |
|   |  |

#### Comment Moderation Allowed

1. Run Jetpack and login.
2. Navigate to "Notifications" screen.
3. Select a "Comment" notification.
4. Compare the implementation against the [design](https://www.figma.com/file/yWt5gg3nWORhu079Qfv3mS/%F0%9F%94%94-Notifications?type=design&node-id=1120-8764&mode=design&t=fPdvnHLLN5zWscDH-4).
5. If the comment has links or mentions, tap one of them and expect a web view is shown.
6. Tap the more menu.
7. Expect these buttons to be visible: "User Info", "Share", "Edit Comment" and "Change Status".
8. Verify "User Info", "Share" and "Edit Comment" buttons work as expected.

#### Comment Moderation Not Allowed

1. Go to Comment.swift, and force allowsModeration to return false.
2. Build and run the app.
3. Navigate to My Site > Comments.
4. Tap any comment.
5. Tap the more menu.
6. Expect these buttons to be visible: "User Info", "Share".

## Regression Notes
1. Potential unintended areas of impact
Smoke test the "Comment Detail" screen when accessed from "My Site > Comments".

3. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

4. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Testing checklist:

- [x] WordPress.com sites and self-hosted Jetpack sites.
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] VoiceOver.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [x] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/ViewRelated/Comments/Content/CommentDetailContentView.swift [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import UIKit
+
+final class CommentDetailContentView: UIView {
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Comments/Content/CommentDetailContentView.swift:3]: As I've mentioned in the PR description, most of the logic was extracted from `CommentContentTableViewCell`, except:
- The cell was using a `NSLayoutConstraint` to constraint the height, I'm using `intrinsicContentSize` instead.
- The cell was providing 2 methods to render comments ( `web`, `richContent` ). I've extracted only the `web` rendering method because it was the one used in `Comment Detail` screen.
- Added a `Configuration` type to encapsulate configuration properties.

File: WordPress/Classes/ViewRelated/Comments/Content/CommentDetailContentView.swift [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        case loading
+        case rendered(CGFloat)
+
+        var height: CGFloat {
+            switch self {
+            case .loading: return .leastNormalMagnitude
+            case .rendered(let height): return height
+            }
+        }
+    }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Comments/Content/CommentDetailContentView.swift:69]: This `State` enum is not very useful right now, but I was thinking of showing an activity indicator because the comment rendering is async.

File: WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift [hunk 3]
```diff
@@ -188,29 +188,6 @@ class CommentDetailViewController: UIViewController, NoResultsViewHost {
...
-            image: comment.allowsModeration()
-            ? UIImage(systemName: Style.Content.ellipsisIconImageName)
-            : UIImage(systemName: Style.Content.shareIconImageName),
-            style: .plain,
-            target: self,
-            action: #selector(shareCommentURL)
-        )
-        button.accessibilityLabel = NSLocalizedString("Share comment", comment: "Accessibility label for button to share a comment from a notification")
-        return button
-    }()
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift:212]: Edit and share bar buttons were removed from the navigation bar, because now they're buttons in the comment's menu.

File: WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift [hunk 4]
```diff
@@ -471,29 +437,32 @@ private extension CommentDetailViewController {
...
+            }
+            switch option {
+            case .userInfo:
+                presentUserInfoSheet()
+            case .editComment:
+                editButtonTapped()
+            case .share:
+                shareCommentURL()
+            case .changeStatus(let status):
+                print("Option \(status) tapped")
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift:453]: Handling status changes will be addressed in a separate PR as it depends on this [issue](https://github.com/Automattic/wordpress-mobile/issues/33).

File: WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift [hunk 5]
```diff
@@ -471,29 +437,32 @@ private extension CommentDetailViewController {
...
+            }
         }
-
-        cell.contentLinkTapAction = { [weak self] url in
+        let contentConfig = CommentDetailContentTableViewCell.ContentConfiguration(comment: comment) { [weak self] _ in
+            UIView.performWithoutAnimation {
+                self?.tableView.performBatchUpdates({})
+            }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift:459]: When rendering the comment, we use the `performBatchUpdates` method to relayout the cells and update their heights. This approach was part of the previous implementation. What I've added is the wrapping of this method inside `UIView.performWithoutAnimation`, ensuring that the cells' relayout occurs without any animation. In my opinion, the default animation appeared clunky.

File: WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift [hunk 6]
```diff
@@ -729,7 +698,7 @@ private extension CommentDetailViewController {
         present(activityViewController, animated: true, completion: nil)
     }
 
-    func presentUserInfoSheet(_ senderView: UIView) {
+    func presentUserInfoSheet() {
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift:701]: The `senderView: UIView` param was not used, so I removed it.

File: WordPress/Classes/ViewRelated/Comments/Content/CommentDetailContentTableViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        self.contentView.pinSubviewToAllEdgeMargins(stackView)
+        self.selectionStyle = .none
+    }
+
+    // MARK: - Updating UI
+
+    func configure(with headerConfig: HeaderConfiguration, contentConfig: ContentConfiguration, parent: UIViewController) {
+        self.updateHeader(with: headerConfig, parent: parent)
+        self.commentView.configure(with: contentConfig)
+    }
```
Comment (justtwago) [inline WordPress/Classes/ViewRelated/Comments/Content/CommentDetailContentTableViewCell.swift:51]: Just to double-check: what is the purpose of this initializer? It is not used and I'm not sure if we have a case for this.

**Review by justtwago — COMMENTED** [review]
Tested the changes and everything works well. My only concern is a really little padding between a header and content. It looks really tight. Maybe it's worth double check with Chris and perhaps restore the previous value?

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/ViewRelated/Comments/Content/CommentDetailContentTableViewCell.swift (hunk 7 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Comments/Content/CommentDetailContentTableViewCell.swift:51]: This acts as the primary `configure` method, while the other one is merely a convenience method. The convenience configure method is supposed to call this one, but I just realized that it's not the case. Resolved in 3d77d0a568c985c2bc3fb1515218c2d3df8cc9b5
